python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py

```diff
@@ -212,42 +223,77 @@ def _get_component_type(component: Component) -> ComponentType:
     to infer the component type.
     """
     component_name = _get_component_class_name(component)
-    if (run_method := getattr(component, "run", None)) is None or not callable(run_method):
+    if (run_method := _get_component_run_method(component)) is None:
         return ComponentType.UNKNOWN
     if "Generator" in component_name or "VertexAIImage" in component_name:
         return ComponentType.GENERATOR
     elif "Embedder" in component_name:
         return ComponentType.EMBEDDER
-    elif "Retriever" in component_name or _has_retriever_run_method(run_method):
+    elif "Ranker" in component_name or _has_ranker_io_types(run_method):
+        return ComponentType.RANKER
+    elif "Retriever" in component_name or _has_retriever_io_types(run_method):
         return ComponentType.RETRIEVER
     elif isinstance(component, PromptBuilder):
         return ComponentType.PROMPT_BUILDER
     return ComponentType.UNKNOWN
 
 
-def _has_retriever_run_method(run_method: Callable[..., Any]) -> bool:
+def _get_component_run_method(component: Component) -> Optional[Callable[..., Any]]:
+    """
+    Gets the `run` method for a component (if one exists).
+    """
+    if callable(run_method := getattr(component, "run", None)):
+        return cast(Callable[..., Any], run_method)
+    return None
+
+
+def _get_run_method_output_types(run_method: Callable[..., Any]) -> Optional[Dict[str, type]]:
+    """
+    Haystack components are decorated with an `output_type` decorator that is
+    useful for inferring the component type.
+
+    https://github.com/deepset-ai/haystack/blob/21c507331c98c76aed88cd8046373dfa2a3590e7/haystack/core/component/component.py#L398
+    """
+
+    if isinstance((output_types_cache := getattr(run_method, "_output_types_cache", None)), dict):
+        return {key: value.type for key, value in output_types_cache.items()}
+    return None
+
+
+def _get_run_method_input_types(run_method: Callable[..., Any]) -> Optional[Dict[str, type]]:
+    """
+    Gets input types of parameters to the `run` method.
+    """
+    return get_type_hints(run_method)
+
+
+def _has_ranker_io_types(run_method: Callable[..., Any]) -> bool:
+    """
+    Uses heuristics to infer if a component has a ranker-like `run` method.
+    """
+    if (input_types := _get_run_method_input_types(run_method)) is None or (
+        output_types := _get_run_method_output_types(run_method)
+    ) is None:
+        return False
+    has_documents_parameter = input_types.get("documents") is List[Document]
```

Comment:
how does this work: `is List[Document]`?

If i do this below I get False:
```python
[1,2] is List[int]
```

Reply:
Checking that the type is the same. So `List[int] == List[int]`.

Reply:
oh, then `is` is the customary choice